### PR TITLE
[RISCV] Fix coalesced vsetvli's AVL LiveInterval not always being shrunk

### PR DESCRIPTION
**Pull request by lukel97**
Most of the time when we coalesce and delete a vsetvli, we shrink the LiveInterval of its AVL register now that there is one less use. However there's one edge case we were missing where if we have two vsetvlis with no users of vl or vtype in between, we coalesced a vsetvli without shrinking it's AVL.

This fixes it by shrinking the LiveIntervals whenever we delete a vsetvli, and also makes the LiveIntervals consistent in-situ by not removing the use before shrinking.

This fixes a -verify-machineinstrs assertion in an MIR test case I found while investigating https://github.com/llvm/llvm-project/pull/97264#issuecomment-2218036877. I couldn't recreate this at the LLVM IR level, seemingly because RISCVInsertVSETVLI will just avoid inserting extra vsetvlis that don't need coalesced.

### CODE REVIEW

File: llvm/lib/Target/RISCV/RISCVInsertVSETVLI.cpp [hunk 1]
```diff
@@ -1727,6 +1727,10 @@ void RISCVInsertVSETVLI::coalesceVSETVLIs(MachineBasicBlock &MBB) const {
     if (LIS)
       LIS->RemoveMachineInstrFromMaps(*MI);
     MI->eraseFromParent();
+    if (LIS)
+      for (MachineOperand &MO : MI->uses())
```
Comment (preames) [inline llvm/lib/Target/RISCV/RISCVInsertVSETVLI.cpp:1731]: Calling MI->uses() after MI->eraseFromParent() is definitely unsafe.

**Review by preames — CHANGES_REQUESTED** [review]

**Review by preames — COMMENTED** [review]
I realized when taking a deeper look that you're duplicating functionality here.  This same case is supposed to be handled in the "if (DefReg.isVirtual() && LIS) {" above.  I think what happens is that you've got an order of operations bug.  You need to *actually* modify the instruction first, then perform all of the LIS update.  (That is, move that block done to just before the OldVReg update, and maybe move the setDesc call up.

**Comment by lukel97** [comment]
DefReg is a different register, it's the output VL, this is for the input AVL. But now that you mention it, we were also duplicating the liveintervals update for the AVL too. We already handle it for the normal case, but the bug was ultimately caused by this case here where we continue early for two consecutive vsetvlis with no pseudos in between:

https://github.com/llvm/llvm-project/pull/98286/files#diff-a186cfa1b7676c0a95a09935007221c321cfc52ce54b85d15064963d413725a3R1657-R1661

I've removed that redundant case so we're only updating the *AVL* in one place.

**Comment by preames** [comment]
Sorry for the confusion, you're right about DefReg vs OldVLReg.  I think you're also right about the early continue being the trigger, but looking at that code, it looks like we're also missing the ADDI prune in that case as well.  Maybe pull out a lambda for prepareToDelete which does all the LIS fixup, ADDI deletion, and addition to the ToDelete list?

**Review by preames — APPROVED** [review]
LGTM to the current patch.

I talked with Luke offline, and was convinced this is correct as structured.  I'm not a fan of the style - doing the updates in two places, and out of order - but I'm explicitly okay with this landing, and I will post an NFC tomorrow.

**Pull request merged**